### ISSUE TITLE
MA plotting error

### ISSUE THREAD

**Issue by beroult** [reskejak/ATAC-seq#3]
Hi, 

In the csaw codes, I get such an error below while plotting the MA plot. The code seems to subtract the logCPM data while selecting the best model which it selected for logFC and I am not sure how to plot it. Thanks in advance for the help. 

ggplot(data=data.frame(Stage1_final.merged.peaks),
       aes(x = logCPM, y = logFC, col = factor(sig, levels=c("n.s.", "significant")))) + 
  geom_point() + scale_color_manual(values = c("black", "red")) + 
  geom_smooth(inherit.aes=F, aes(x = logCPM, y = logFC), method = "loess") + 
  geom_hline(yintercept = 0) + labs(col = NULL)

Error in FUN(X[[i]], ...) : object 'logCPM' not found

**Comment by AvitalPo** [comment]
Hi,
I have the same issue, how it was resolved?

Thanks

**Comment by beroult** [comment]
I thought I resolved the issue but actually did not. So reopened it. Please share your help if you can. Thanks

**Comment by reskejak** [comment]
My apologies for the delay.

Can you please provide the output of `head(data.frame(Stage1_final.merged.peaks))`? One of the `colnames` is not corresponding to `logCPM`. Try `colnames(data.frame(Stage1_final.merged.peaks))` as well.

Thanks, Jake

**Comment by beroult** [comment]
Hi, 

`head(data.frame(Stage1_final.merged.peaks))`

> seqnames start   end width strand num.up.logFC num.down.logFC       PValue          FDR direction rep.test rep.logFC         sig
1     S811 43107 43880   774      *            0              1 5.855526e-12 9.027465e-08      down    14196 -6.641337 significant
2     S455   392   928   537      *            1              0 1.821894e-11 1.404407e-07        up     9537  6.866849 significant
3     S586 67066 67820   755      *            0              1 5.312421e-11 2.047540e-07      down    11556 -2.626026 significant
4      S74 58671 58988   318      *            1              0 5.267126e-11 2.047540e-07        up    13304  5.077049 significant
5     S518  1949  2963  1015      *            1              0 2.883129e-10 7.408200e-07        up    10581  1.760364 significant
6     S586 57491 61289  3799      *            0              2 2.549561e-10 7.408200e-07      down    11551 -1.208434 significant

`colnames(data.frame(Stage1_final.merged.peaks))`

>  [1] "seqnames"       "start"          "end"            "width"          "strand"         "num.up.logFC"   "num.down.logFC" "PValue"         "FDR"            "direction"      "rep.test"       "rep.logFC"      "sig"           


Indeed I don't see any logCPM calculated . 
Thanks in advance. 
Beri

**Comment by reskejak** [comment]
I am assuming there were updates to the csaw code that affect the columns of the resulting tables. This step is probably causing the issue:
```
# concatenating all relevant statistical data for final merged windows (no redundant columns)
final.merged.peaks <- GRanges(cbind(as.data.frame(merged.peaks$region), results$table[tab.best$rep.test, -4], tab.best[,-c(7:8)]))
```

Please send along `head(as.data.frame(merged.peaks$region))` and `head(results$table[tab.best,])` and `head(tab.best)`.

**Comment by beroult** [comment]
`head(as.data.frame(Stage1_final.merged.peaks$region))`

> data frame with 0 columns and 0 rows


`head(results1$table[Stage1_tab.best,])`

> Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'x' in selecting a method for function 'head': invalid subscript type 'S4'

`head(Stage1_tab.best)`

> DataFrame with 6 rows and 8 columns
  num.tests num.up.logFC num.down.logFC    PValue       FDR   direction  rep.test  rep.logFC
  <integer>    <integer>      <integer> <numeric> <numeric> <character> <integer>  <numeric>
1         1            0              0 0.5542249  0.804805        down         1 -0.1349076
2         1            0              0 0.0647421  0.349241        down         2 -0.6475463
3         1            0              0 0.2572499  0.601095        down         3 -0.2621583
4         1            0              0 0.1213251  0.449632          up         4  0.3901992
5         1            1              0 0.0296041  0.249847          up         5  0.3530079
6         1            0              0 0.9156599  0.976173        down         6 -0.0212544

**Comment by reskejak** [comment]
For `head(as.data.frame(merged.peaks$region))`, this is the peak set before you have combined everything into one "final" table. I am trying to figure out what is happening at the step listed in the previous comment. In your code, it looks like this should NOT be `Stage1_final.merged.peaks`, perhaps rather `Stage1_merged.peaks`?

**Comment by beroult** [comment]
Sorry for the confusion. See below. 

`head(as.data.frame(Stage1_merged.peaks$region))`

>   seqnames start   end width strand
1       S0   519   750   232      *
2       S0  7524  7672   149      *
3       S0  9986 10888   903      *
4       S0 28049 28293   245      *
5       S0 37301 38024   724      *
6       S0 38658 39124   467      *

**Comment by reskejak** [comment]
From what I can tell, there must have been some update to the csaw functions that is causing the desired tables to have different output columns. I'll need to re-run the workflow with the latest csaw version to fix this once I have some time. My apologies!

**Comment by beroult** [comment]
Algood! Thanks for your time and help. Much appreciated.

**Comment by sufyazi** [comment]
Hi I am here because I am facing the exact same issue with this step. It seems like log2CPM values are not retained in the `final_merged_peaks`. Can I know if this has been investigated?

Otherwise, I can probably troubleshoot this myself, but R isn't exactly my forte. I tried tracing down logCPM values in the workflow and the only logCPM values I could find appearing last was in `results$table`. 

In the `# combine GRanges rowdata with DA statistics` step, doing `rowData(working_windows) <- cbind(rowData(working_windows), results$table)` does retain the logCPM column in `@rowRanges` but somehow this gets dropped somewhere on the way to `final_merged_peaks`. I suspect this occurs during the `# merge nearby windows` step, where the `mergeWindows` command returns a large list of 2 ($id and $region). I am not sure how to retain logCPM here.

The next step is `# use the most significant windows as stat representation` with `tab_best <- getBestTest(merged_peaks$id, results$table)`. It is where I think the logCPM is lost. This command seems to not be keeping the logCPM column in its output.

I wonder if we can manually extract the logCPM column from `results$table` and append it to `final_merged_peaks`? I just worry if this mess up with the row data because there is a sorting step (by FDR) to get the `final_merged_peaks`.

I am using `csaw_1.32.0` by the way.

**Comment by reskejak** [comment]
@sufyazi I suspect you are correct here. I will investigate this as soon as possible and get back to you shortly.

**Comment by reskejak** [comment]
@sufyazi I just updated to the latest R and csaw versions and was not able to recapitulate your issue. Everything seems to be working for me.

> I wonder if we can manually extract the logCPM column from results$table and append it to final_merged_peaks?

That is exactly what we are doing during the creation of `final.merged.peaks` with the object `results$table[tab.best$rep.test,]`, which takes the respective indices of `results$table` selected by `getBestTest()`.

As mentioned in https://github.com/reskejak/ATAC-seq/issues/3#issuecomment-1211919829, I suspect the loss of `logCPM` is in:
```
final.merged.peaks <- GRanges(cbind(as.data.frame(merged.peaks$region), results$table[tab.best$rep.test, -4], tab.best[,-c(7:8)]))
```
This is combining a few objects into one GRanges table while discarding some redundant columns. On some random toy data, we have the following:
```
head(as.data.frame(merged.peaks$regions))

  seqnames  start    end width strand
1     chr1 827070 828081  1012      *
2     chr1 904262 907994  3733      *
3     chr1 922600 925554  2955      *
4     chr1 938329 943397  5069      *
5     chr1 959062 961828  2767      *
6     chr1 965934 967140  1207      *

head(results$table[tab.best$rep.test,])

             logFC    logCPM            F       PValue
12293 -0.304979685 0.3847920  2.884651274 0.1016100978
12675 -0.144384360 1.4944314  1.288865587 0.2562617921
13351 -0.146397962 1.3282484  1.185985430 0.2761453531
13809 -0.008597626 2.2227517  0.007349781 0.9316806618
14632 -0.383772875 1.7784576 10.845668698 0.0009908619
14681 -0.268084609 0.6213182  2.570632135 0.1174949625

head(tab.best)

DataFrame with 6 rows and 8 columns
  num.tests num.up.logFC num.down.logFC      PValue       FDR   direction  rep.test   rep.logFC
  <integer>    <integer>      <integer>   <numeric> <numeric> <character> <integer>   <numeric>
1         1            0              0 0.101610098 0.4172062        down     12293 -0.30497968
2         1            0              0 0.256261792 0.6213054        down     12675 -0.14438436
3         1            0              0 0.276145353 0.6425801        down     13351 -0.14639796
4         1            0              0 0.931680662 0.9825565        down     13809 -0.00859763
5         1            0              1 0.000990862 0.0376785        down     14632 -0.38377287
6         1            0              0 0.117494962 0.4469153        down     14681 -0.26808461

# combined into final.merged.peaks
head(final.merged.peaks)

GRanges object with 6 ranges and 9 metadata columns:
        seqnames        ranges strand |       logFC    logCPM           F num.tests num.up.logFC num.down.logFC      PValue       FDR   direction
           <Rle>     <IRanges>  <Rle> |   <numeric> <numeric>   <numeric> <integer>    <integer>      <integer>   <numeric> <numeric> <character>
  12293     chr1 827070-828081      * | -0.30497968  0.384792  2.88465127         1            0              0 0.101610098 0.4172062        down
  12675     chr1 904262-907994      * | -0.14438436  1.494431  1.28886559         1            0              0 0.256261792 0.6213054        down
  13351     chr1 922600-925554      * | -0.14639796  1.328248  1.18598543         1            0              0 0.276145353 0.6425801        down
  13809     chr1 938329-943397      * | -0.00859763  2.222752  0.00734978         1            0              0 0.931680662 0.9825565        down
  14632     chr1 959062-961828      * | -0.38377287  1.778458 10.84566870         1            0              1 0.000990862 0.0376785        down
  14681     chr1 965934-967140      * | -0.26808461  0.621318  2.57063213         1            0              0 0.117494962 0.4469153        down
```

Could you try repeating that above code and post your results? Hope this helps!


```
R version 4.2.3 (2023-03-15)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Ventura 13.2.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] edgeR_3.40.2                limma_3.54.2                csaw_1.32.0                 SummarizedExperiment_1.28.0 Biobase_2.58.0             
 [6] MatrixGenerics_1.10.0       matrixStats_0.63.0          GenomicRanges_1.50.2        GenomeInfoDb_1.34.9         IRanges_2.32.0             
[11] S4Vectors_0.36.2            BiocGenerics_0.44.0        

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.10            XVector_0.38.0         splines_4.2.3          zlibbioc_1.44.0        MASS_7.3-58.3          statmod_1.5.0         
 [7] BiocParallel_1.32.6    lattice_0.20-45        tools_4.2.3            parallel_4.2.3         grid_4.2.3             crayon_1.5.2          
[13] Matrix_1.5-3           GenomeInfoDbData_1.2.9 bitops_1.0-7           codetools_0.2-19       RCurl_1.98-1.10        DelayedArray_0.24.0   
[19] compiler_4.2.3         Biostrings_2.66.0      Rsamtools_2.14.0       metapod_1.6.0          locfit_1.5-9.7  
```

**Comment by sufyazi** [comment]
Thank you for helping me out. 

Okay I followed your code above, and the final `final.merged.peaks` object looks correct to me (demonstrated below with my dataset).

```
head(final_merged_peaks_normt)
GRanges object with 6 ranges and 9 metadata columns:
    seqnames        ranges strand |     logFC    logCPM         F num.tests num.up.logFC num.down.logFC    PValue       FDR   direction
       <Rle>     <IRanges>  <Rle> | <numeric> <numeric> <numeric> <integer>    <integer>      <integer> <numeric> <numeric> <character>
  1     chr1 826771-828088      * | -0.484522   4.31220  0.290358         1            0              0  0.589992  0.791019        down
  2     chr1 865489-866015      * | -0.963272   2.22663  1.246546         1            0              0  0.264214  0.597150        down
  3     chr1 869732-870247      * | -1.060550   1.68189  1.596431         1            0              0  0.206411  0.529303        down
  5     chr1 904412-905598      * | -0.868416   1.01465  1.171702         2            0              0  0.558105  0.776948        down
  6     chr1 906782-907144      * | -0.264133   1.02258  0.105253         1            0              0  0.745615  0.867827        down
  7     chr1 959000-959530      * |  0.276608   3.00823  0.093460         1            0              0  0.759824  0.875890          up
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths

```

This step makes more sense in terms of analysis logic to me now, but these steps are not present in the original source code I obtained from the supplementary of the associated publication of yours (Reske et al. 2020), which should explain reader's confusion when trying to replicate this analysis workflow.

Verbatim, a snippet from the final part of the R script:

```
# merge nearby windows
# up to "tol" distance apart: 500 bp in this case
# max merged window width: 5000 bp
merged.peaks <- mergeWindows(rowRanges(working.windows), tol=500L, max.width=5000L)
# summary(width(merged.peaks$region))
# should merge some peaks; change as desired

# use most significant window as statistical representation for p-value and FDR for merged windows
tab.best <- getBestTest(merged.peaks$id, results$table)
head(tab.best)

# combine merged peaks window range with statistics
final.merged.peaks <- merged.peaks$region
final.merged.peaks@elementMetadata <- cbind(final.merged.peaks@elementMetadata, tab.best[,-1])
final.merged.peaks <- final.merged.peaks[order(final.merged.peaks@elementMetadata$FDR), ] # sort by FDR
final.merged.peaks # all windows

# filter by FDR threshold
FDR.thresh <- 0.05 # set as desired
final.merged.peaks.sig <- final.merged.peaks[final.merged.peaks@elementMetadata$FDR < FDR.thresh, ]
final.merged.peaks.sig # significant differentially-accessible windows

write.table(final.merged.peaks, "treat_vs_control_csaw_DA-windows_all.txt", sep="\t", quote=F, col.names=T, row.names=F)
write.table(final.merged.peaks.sig, "treat_vs_control_csaw_DA-windows_significant.txt", sep="\t", quote=F, col.names=T, row.names=F)

```

This is before the `MA plotting` section. I guess the submitted script was accidentally truncated? 

Either way, my question now is, where does the step above fit into the rest of the workflow? I believe it should be after `tab.best` (obviously), and then the statistics metadata and FDR filtering steps follow right after selecting only the rows in the `results$table` selected by `tab.best$rep.test`? 

When I run the statistics metadata part `# combine merged peaks window range with statistics`, it clearly overrides the command you suggested above, specifically when I ran this `final.merged.peaks <- merged.peaks$region`. I believe the FDR filtering is crucial for the final list, so below is the modified workflow I did for my dataset. 

```
# view only the best windows (rows) from results_normt$table using the rep.test column of tab_best
head(results_normt$table[tab_best_normt$rep.test, ], n=10)

# create final merged peak sets (sorted by FDR metadata column)
final_merged_peaks_normt <- GRanges(cbind(as.data.frame(merged_peaks_normt$regions), results_normt$table[tab_best_normt$rep.test, -4], tab_best_normt[, -c(7:8)]))
final_merged_peaks_normt # all windows

final_merged_peaks_normt <- final_merged_peaks_normt[order(final_merged_peaks_normt@elementMetadata$FDR), ] # sort by FDR
final_merged_peaks_normt # all windows sorted

# filter by FDR threshold
FDR.thresh <- 0.05 # set as desired
final_merged_peaks_normt.sig <- final_merged_peaks_normt[final_merged_peaks_normt@elementMetadata$FDR < FDR.thresh, ]
final_merged_peaks_normt.sig # significant differentially-accessible windows

# write to files
write.table(final_merged_peaks_normt, "csaw_dar_o/mcl_vs_nbctons-c5_csaw_DA-windows_all.txt", sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)
write.table(final_merged_peaks_normt.sig, "csaw_dar_o/mcl_vs_nbctons-c5_csaw_DA-windows_sig.txt", sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

```
Does this revision to the workflow look correct to you? The MA plot was able to be generated after this point, by the way.

I think all in all, if the way I incorporated your fix into the original workflow is correct, then I guess the issue can be considered solved. Thank you again for your time and feel free to close this.

**Comment by reskejak** [comment]
Hi again,

I should have first checked to see if you were using the latest R script in the repo. Unfortunately there was a major csaw update right after publication...

This updated script should include the current csaw commands as we have discussed in this thread:
https://github.com/reskejak/ATAC-seq/blob/master/csaw_workflow.R
Go ahead and check your code to see if it lines up.

Closing this issue for now! Let me know if you have any other problems or questions.
